### DISCUSSION TITLE
Handling errors without exposing messages to the client

### DISCUSSION THREAD

**Discussion by szanuje** [trpc/trpc#3844] [Q&A] [answered]
Hey, let's consider such scenarios:
- My procedure does an internal API call to AWS service. There's a connection issue, so the error is being thrown
- My procedure does Prisma query and an error is being thrown

Now, the client can see the error and its message, which is bad because it can leak the IP address of some backend services.

How can I avoid that?

The only solution that I can think of, is to try..catch every procedure, report errors to e.g Sentry, and based on that error, throw a new one without confidential data. The disadvantage here is that I would have to report errors in each try...catch, instead of the global trpc handler, because otherwise, I will lose all the context (stack trace, real error message).

Is there something that I'm missing here? Can you propose some solution? Thanks.

**Comment by KATT** [comment, marked as answer]
You can make [a custom error formatter](https://trpc.io/docs/error-formatting)


### Example


```ts
const trpc = initTRPC
  .create({
    errorFormatter({ error, shape }) {
      return {
        ...shape,
        message:
          error.code === 'INTERNAL_SERVER_ERROR' && process.env.NODE_ENV === 'production'
            ? 'Something went wrong'
            : shape.message,
        data: {
          ...shape.data,
        },
      };
    },
  });

```

**Reply by szanuje** [reply]
That's awesome, thanks!
